kite-data/kite-data-core/src/main/java/org/kitesdk/data/View.java

```diff
@@ -128,4 +128,28 @@
    * @since 0.16.0
    */
   public URI getUri();
+
+  /**
+   * Returns an {@code Iterable} of non-overlapping {@link View} objects that
+   * partition the underlying {@link Dataset} and cover this {@code View}.
+   *
+   * The returned {@code View} objects are implementation-specific, but should
+   * represent reasonable partitions of the underlying {@code Dataset} based on
+   * its layout.
+   *
+   * The data contained by the union of each {@code View} in the
+   * {@code Iterable} must be a super-set of this {@code View}.
+   *
+   * Note that partitions are actual partitions under which data is stored.
+   * Implementations should omit any {@code View} that is empty.
+   *
+   * @return
+   *      An Iterable of the {@code View} that cover this {@code View}.
+   * @throws IllegalStateException
+   *      If the underlying {@code Dataset} is not partitioned.
+   * @throws UnsupportedOperationException
+   *      If the implementation does not support covering partitions.
+   * @since 1.1.0
+   */
+  public Iterable<PartitionView<E>> getCoveringPartitions();
```

Comment:
I think you're right that this should require that each view returned overlaps part of the original view and could have data in it. That's definitely an oversight.

As for what should be returned by implementations, I'm not sure what the right thing to do is. The use case I implemented this for needs to operate on the smallest partition level to avoid removing partitions that are peers of partitions that are being replaced (the 31-element iterable in your example). We could change that, but I'm not sure which is more correct.

One consideration: the root will be the minimal partition in a lot of cases, when you have no partition constraints. In that case, it's misleading to think you can simply move the directories because there are .metadata and .signals folders that would be lost. So maybe the right thing to do is to return all of the full paths as partition views and allow other FS code to analyze what's on disk to merge those into work units. But then, that puts a lot of work to get efficient operations on the user so I'm not too happy either way.

Thoughts?